src/scope.rs

```diff
@@ -539,1771 +1224,877 @@ impl<P> TryCatch<'_, P> {
-    /// because the 'a lifetime ends there.
-    ///
-    /// Scope types that do no store local handles are exited immediately.
-    #[inline(always)]
-    pub(super) fn notify_scope_dropped(&mut self) {
-      match &self.scope_type_specific_data {
-        ScopeTypeSpecificData::HandleScope {
-          allow_zombie: true, ..
-        }
-        | ScopeTypeSpecificData::EscapableHandleScope { .. } => {
-          // Defer scope exit until the parent scope is touched.
-          self.status.set(match self.status.get() {
-            ScopeStatus::Current { zombie: false } => {
-              ScopeStatus::Current { zombie: true }
-            }
-            _ => unreachable!(),
-          });
-        }
-        _ => {
-          // Regular, immediate exit.
-          self.exit_scope();
-        }
-      }
-    }
+pub trait NewAllowJavascriptExecutionScope<'scope> {
+  type NewScope: Scope;
+  fn make_new_scope(me: &'scope mut Self) -> Self::NewScope;
+}
 
-    #[inline(always)]
-    pub(crate) fn get_isolate(&self) -> &Isolate {
-      unsafe { self.isolate.as_ref() }
+impl<'scope, 'obj: 'scope, P: Scope + GetIsolate>
+  NewAllowJavascriptExecutionScope<'scope> for PinnedRef<'obj, P>
+{
+  type NewScope = AllowJavascriptExecutionScope<'scope, 'obj, P>;
+  fn make_new_scope(me: &'scope mut Self) -> Self::NewScope {
+    AllowJavascriptExecutionScope {
+      raw: unsafe { raw::AllowJavascriptExecutionScope::uninit() },
+      scope: me,
+      _pinned: PhantomPinned,
     }
+  }
+}
 
-    #[inline(always)]
-    pub(crate) fn get_isolate_mut(&mut self) -> &mut Isolate {
-      unsafe { self.isolate.as_mut() }
-    }
+// Note: the macros below _do not_ use std::pin::pin! because
+// it leads to worse compiler errors when the scope doesn't live long enough.
+// Instead, we do the same thing as std::pin::pin! but without the additional temporary scope.
+
+#[allow(unused_macros, clippy::macro_metavars_in_unsafe)]
+#[macro_export]
+macro_rules! make_callback_scope {
+  (unsafe $scope: ident, $param: expr) => {
+    #[allow(clippy::macro_metavars_in_unsafe)]
+    let mut $scope = {
+      // force the caller to put a separate unsafe block around the param expr
+      let param = $param;
+      unsafe { $crate::CallbackScope::new(param) }
+    };
+    // SAFETY: we are shadowing the original binding, so it can't be accessed
+    // ever again
+    let mut $scope = {
+      let scope_pinned = unsafe { std::pin::Pin::new_unchecked(&mut $scope) };
+      scope_pinned.init()
+    };
+    let $scope = &mut $scope;
+  };
+  (unsafe let $scope: ident, $param: expr) => {
+    $crate::make_callback_scope!(unsafe $scope, $param);
+  }
+}
 
-    #[inline(always)]
-    pub(crate) fn get_isolate_ptr(&self) -> *mut Isolate {
-      self.isolate.as_ptr()
-    }
+#[allow(unused_imports)]
+pub(crate) use make_callback_scope;
+
+#[allow(unused_macros)]
+#[macro_export]
+macro_rules! make_handle_scope {
```

Comment:
```suggestion
/// Creates a pinned `HandleScope` and binds `&mut PinScope` to `$scope`.
/// 
/// ```
/// v8::make_handle_scope!(let scope, isolate);
/// ```
#[allow(unused_macros)]
#[macro_export]
macro_rules! make_handle_scope {
```